### PR TITLE
Tweaks to reduce memory use of short commands

### PR DESCRIPTION
**Pull request by headius**
This is a set of memory/alloc-reducing experiments that can be merged or cherry-picked as appropriate.

### CODE REVIEW

**Comment by headius** [comment]
The failures appear to be due to 3dc4a9c, the first commit in this PR. For some reason the line number is off by one. @enebo Can you see why that would be?

**Comment by headius** [comment]
I have re-pushed this with the failing source position patch moved to #5838.

File: core/src/main/java/org/jruby/ast/ListNode.java [hunk 1]
```diff
@@ -174,11 +246,16 @@ public boolean isEmpty() {
...
+        Node single = this.single;
+
+        if (idx == 0 && single != null) return single;
+
         return list[idx];
     }
 
     @Override
     public Iterator<Node> iterator() {
+        Thread.dumpStack();
```
Comment (kares) [inline core/src/main/java/org/jruby/ast/ListNode.java]: :scissors:

Comment (headius) [reply core/src/main/java/org/jruby/ast/ListNode.java]: Fixed! We might want to consider removing this since it doesn't appear to be used.

**Comment by headius** [comment]
I'm going to go ahead and merge this. @enebo any review issues you find I will patch later.

**Pull request merged**